kernel/arch/dreamcast/include/dc/maple/keyboard.h

```diff
@@ -28,283 +40,647 @@ __BEGIN_DECLS
+
+    This macro is used to pack two frames worth of key state information
+    into a key_state_t, one bit per frame.
+*/
+#define KEY_STATE_PACK(is_down, was_down) \
+    (((!!(is_down))?  KEY_STATE_IS_DOWN  : 0) | \
+     ((!!(was_down))? KEY_STATE_WAS_DOWN : 0))
+
+/** \brief Valid values for key_state_t::value
+
+    Enumerates each of the 4 different states a key can be in,
+    by combining two frames worth of key down information
+    into two bits.
+
+    \note
+    Two of the values are for `HELD` states, meaning the same state has been
+    observed for both the current and the previous frame, while the other two
+    values are for `CHANGE` states, meaning the current frame has a different
+    state from the previous frame.
+
+    \sa key_state_flags, key_state_t::value
+*/
+typedef enum __packed key_state_value {
+    /** \brief Key has been in an up state for at least the last two frames */
+    KEY_STATE_HELD_UP      = KEY_STATE_PACK(false, false),
+    /** \brief Key transitioned from up to pressed this frame */
+    KEY_STATE_CHANGED_DOWN = KEY_STATE_PACK(true, false),
+    /** \brief Key transitioned from down to released this frame */
+    KEY_STATE_CHANGED_UP   = KEY_STATE_PACK(false, true),
+    /** \brief Key has been held down for at least the last two frames */
+    KEY_STATE_HELD_DOWN    = KEY_STATE_PACK(true, true),
+} key_state_value_t;
+
+/** \brief Keyboard Key State
+
+    Union containing the the previous and current frames' state information for
+    a keyboard key.
+
+    \sa key_state_flags, key_state_value_t, kbd_state::key_states
+*/
+typedef union key_state {
+    /** \brief Convenience Bitfields */
+    struct {
+        uint8_t is_down  : 1; /**< \brief Whether down the current frame */
+        uint8_t was_down : 1; /**< \brief Whether down the previous frame */
+        uint8_t          : 6;
+    };
+    key_state_value_t value;  /**< \brief Enum for specific state */
+    uint8_t           raw;    /**< \brief Packed 8-bit unsigned integer of bitflags */
+} key_state_t;
+
+/** @} */
+
+/** \brief      Region Codes for the Dreamcast keyboard
+    \ingroup    kbd_status_grp
+
+    This is the list of possible values for kbd_state_t::region.
+*/
+typedef enum kbd_region {
+    KBD_REGION_JP = 1, /**< \brief Japanese keyboard */
+    KBD_REGION_US = 2, /**< \brief US keyboard */
+    KBD_REGION_UK = 3, /**< \brief UK keyboard */
+    KBD_REGION_DE = 4, /**< \brief German keyboard */
+    KBD_REGION_FR = 5, /**< \brief French keyboard */
+    KBD_REGION_IT = 6, /**< \brief Italian keyboard (not supported yet) */
+    KBD_REGION_ES = 7  /**< \brief Spanish keyboard */
+} kbd_region_t;
 
 /** \brief   Maximum number of keys a DC keyboard can have.
-    \ingroup kbd
-    This is a hardware constant. The define prevents the magic number '256' from appearing.
+
+    This is a hardware constant. The define prevents the magic number '256'
+    from appearing.
 **/
-#define MAX_KBD_KEYS 256
```

Comment:
It has to stay here in the header, because we're actually using it within the public segment of the structure for the key state array size at line `312`.